File: level2/ex58.py
from collections import deque

def solution(maps):
    # n x m 지도
    n, m = len(maps), len(maps[0])
    # 큐 for bfs
    q = deque()
    q.append((0, 0, 1))
    # 방문 좌표
    visited = set([(0, 0)])
    # 상하좌우 이동량
    dx = [0, 0, -1, 1]
    dy = [-1, 1, 0, 0]
    # (0, 0)으로부터의 최단 거리
    FAILURE = 10000
    distance = [[FAILURE] * m for _ in range(n)]
    
    while q:
        # 탐색 좌표
        x, y, count = q.popleft()
        distance[x][y] = min(distance[x][y], count)
        # 상하좌우 확인
        for i in range(4):
            nx, ny = x+dx[i], y+dy[i]
            # 범위 이탈
            if nx not in range(n) or ny not in range(m):
                continue
            # 방문한 경우
            if (nx, ny) in visited:
                continue
            # 최단 경로가 아닌 경우
            if count >= distance[nx][ny]:
                continue
            # 경로 이동
            if maps[nx][ny] == 1:
                q.append((nx, ny, distance[x][y]+1))
                visited.add((nx, ny))
    
    if distance[n-1][m-1] == FAILURE:
        return -1
    else:
        return distance[n-1][m-1]

File: level2/ex58.js
function solution(maps) {
    // n x m 지도
    const n = maps.length;
    const m = maps[0].length;
    // 큐 for bfs
    const q = [];
    q.push([0, 0, 1]);
    // 방문 좌표
    const visited = new Set();
    visited.add([0, 0].join(','));
    // 상하좌우 이동량
    const dx = [0, 0, -1, 1];
    const dy = [-1, 1, 0, 0];
    // (0, 0)으로부터의 최단 거리
    const FAILURE = 10000;
    const distance = Array.from({ length: n}, () => new Array(m).fill(FAILURE));

    while (q.length > 0) {
        // 탐색 좌표
        const [x, y, count] = q.shift();
        distance[x][y] = Math.min(distance[x][y], count);
        
        // 상하좌우 확인
        for (let i = 0; i < 4; i++) {
            const nx = x + dx[i];
            const ny = y + dy[i];
            // 범위 이탈
            if (nx < 0 || nx >= n || ny < 0 || ny >= m) {
                continue;
            } 
            // 방문한 경우
            if (visited.has([nx, ny].join(','))) {
                continue;
            }
            // 최단 경로가 아닌 경우
            if (count >= distance[nx][ny]) {
                continue;
            }
            // 경로 이동
            if (maps[nx][ny] === 1) {
                q.push([nx, ny, distance[x][y]+1]);
                visited.add([nx, ny].join(','));
            }
        }
    }

    if (distance[n-1][m-1] === FAILURE) {
        return -1;
    } else {
        return distance[n-1][m-1];
    }
}